subsys/net/ip/tcp.c

```diff
@@ -417,7 +417,9 @@ int net_tcp_prepare_segment(struct net_tcp *tcp, u8_t flags,
 
 	if (flags & NET_TCP_FIN) {
 		tcp->flags |= NET_TCP_FINAL_SENT;
-		seq++;
+printk("Making sure FIN goes with ACK; adjusting seq by 2; now fire up Wireshark and check that all's ok\n");
+		flags |= NET_TCP_ACK;
+		seq += 2;
```

Comment:
Yes, likely the sequence should have been incremented in some other state, but it wasn't, that's why we need to increment it by 2 here. It would take me some good time to trace thru Z TCP state machine implementation (nor I know TCP state machine itself by heart and will need to refresh myself on it). So I just hope that folks who know more about Z stack can look at this patch and get an idea what may be the underlying cause. But I'm glad that we have in store an option to merge this, if no better fix surfaces ;-).